### PR TITLE
[REF] start to refactor code WIP

### PR DESCRIPTION
**Pull request by sebastienbeau**
Big bang of refactoring : https://github.com/akretion/storage/issues/26

TODO list
- [x] use component for adapter
- [x] split image and thumbnail
- [x] add test
- [x] always set an url to storage.file even if served by odoo
- [x] remove useless code
- [x] add ir.config.parameter for storing image and thumbnail in different backend
- [x] remove res_model and res_id on storage.file
- [x] add storage strategie (based on hash, based on name + id...)
- [x] split storage backend amazon s3
- [x] split storage backend sftp
- [x] add deletation file support
- [x] review access right
- [x] improve code when adding an image (mapping between image_medium_url and datas)
- [x] fix issue with invalid cache after generating the thumbnails

### CODE REVIEW

File: storage_backend/models/storage_backend.py [hunk 1]
```diff
@@ -2,48 +2,51 @@
...
-    backend_type = fields.Selection([], required=True)  # added by subclasses
-    public_base_url = fields.Char()
-
-    def store(self, name, datas, is_base64=True, **kwargs):
+    backend_type = fields.Selection(
+        selection=[
+            ('filesystem', 'Filesystem'),
+            ],
+        required=True)
+    directory_path = fields.Char(
```
Comment (hparfr) [inline storage_backend/models/storage_backend.py:23]: should be put elsewhere

File: storage_backend/tests/test_filesystem.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        self.backend = self.env.ref('storage_backend.default_storage_backend')
+
+    def _get_filepath(self, filename):
+        return os.path.join(
+            config.filestore(self.cr.dbname),
+            'storage',
+            self.backend.directory_path or '',
+            filename)
+
+    def test_00_setting_and_getting_data_from_root(self):
```
Comment (hparfr) [inline storage_backend/tests/test_filesystem.py]: why numbers in function name ?

File: storage_backend/tests/test_filesystem.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from .common import Common
+from odoo.tools import config
+import os
+import base64
+
+
+class FileStoreCase(Common):
```
Comment (hparfr) [inline storage_backend/tests/test_filesystem.py]: FileSystem

File: storage_backend_s3/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Akretion (http://www.akretion.com).
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Storage Backend S3",
+    "summary": "Implement amazon S3 Storage",
+    "version": "10.0.1.0.0",
+    "category": "Uncategorized",
```
Comment (hparfr) [inline storage_backend_s3/__manifest__.py]: category: storage

File: storage_backend_s3/components/s3_adapter.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    _usage = 'amazon_s3'
+
+    def _get_resource(self):
+        account = self.collection._get_existing_keychain()
+        return boto3.Session(
+            aws_access_key_id=self.collection.aws_access_key_id,
+            aws_secret_access_key=account._get_password(),
+            region_name=self.collection.aws_region).resource('s3')
+
+    def _get_amazon_s3_object(self, relative_path):
```
Comment (hparfr) [inline storage_backend_s3/components/s3_adapter.py]: _get_object instead ?

File: storage_backend_s3/models/storage_backend.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    def _get_aws_region(self):
+        session = boto3.session.Session()
+        return [
+            (region, region.replace('-', ' ').capitalize())
+            for region in session.get_available_regions('s3')]
+
+    backend_type = fields.Selection(
+        selection_add=[('amazon_s3', 'Amazon S3')])
+    aws_bucket = fields.Char(
```
Comment (hparfr) [inline storage_backend_s3/models/storage_backend.py:27]: why prefix with aws ?

File: storage_backend_sftp/components/sftp_adapter.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+except ImportError as err:
+    logger.debug(err)
+
+
+def sftp_mkdirs(client, path, mode=511):
+    try:
+        client.mkdir(path, mode)
+    except IOError, e:
+        if e.errno == errno.ENOENT and path:
+            sftp_mkdirs(client, os.path.dirname(path), mode=mode)
```
Comment (hparfr) [inline storage_backend_sftp/components/sftp_adapter.py:27]: why ?

File: storage_file/models/storage_file.py [hunk 8]
```diff
@@ -82,49 +77,59 @@ def write(self, vals):
...
-                size = size / 1024.0
-            record.human_file_size = "%.*f%s" % (
-                2, size, suffixes[suffix_index])
+            record.human_file_size = human_size(self.file_size)
+
+    def _build_relative_path(self, checksum):
+        self.ensure_one()
+        if not self.backend_id.filename_strategy:
+            raise UserError(_(
+                "The filename strategy is empty for the backend %s.\n"
```
Comment (hparfr) [inline storage_file/models/storage_file.py]: s/empty/missing

File: storage_file/models/storage_file.py [hunk 9]
```diff
@@ -82,49 +77,59 @@ def write(self, vals):
...
+                "Please configure it") % self.backend_id.name)
```
Comment (hparfr) [inline storage_file/models/storage_file.py]: No need to "please configure it" or tell where it should be done

File: storage_file/models/storage_file.py [hunk 10]
```diff
@@ -82,49 +77,59 @@ def write(self, vals):
...
+        if self.backend_id.filename_strategy == 'hash':
+            return checksum[:2] + '/' + checksum
```
Comment (hparfr) [inline storage_file/models/storage_file.py]: why?

File: storage_file/models/storage_file.py [hunk 11]
```diff
@@ -82,49 +77,59 @@ def write(self, vals):
...
+        elif self.backend_id.filename_strategy == 'name_with_id':
+            return "%s-%s%s" % (self.filename, self.id, self.extension)
 
-    def _prepare_meta_for_file(self, datas, private_path):
+    def _prepare_meta_for_file(self):
+        bin_data = base64.b64decode(self.datas)
```
Comment (hparfr) [inline storage_file/models/storage_file.py]: why datas is already encoded ?

File: storage_file/models/storage_file.py [hunk 12]
```diff
@@ -82,49 +77,59 @@ def write(self, vals):
...
+                    base_url + '/web/content/storage.file/%s/datas' % self.id
             else:
-                try:
-                    rec.datas = rec.backend_id.sudo().retrieve_data(
-                        rec.private_path)
-                except:
-                    _logger.error('Image %s not found', rec.url)
-                    rec.datas = None
-                    raise
+                record.url = "%s/%s" % (
```
Comment (hparfr) [inline storage_file/models/storage_file.py]: the construction of the url should be delegated to the backend.
You can't assume here any pattern.

File: storage_image/models/storage_image.py [hunk 13]
```diff
@@ -19,27 +19,15 @@
 class StorageImage(models.Model):
     _name = 'storage.image'
     _description = 'Storage Image'
+    _inherit = 'thumbnail.owner'
```
Comment (hparfr) [inline storage_image/models/storage_image.py]: why keep owner ?

File: storage_image/static/src/js/storage_image.js [hunk 14]
```diff
@@ -15,23 +13,19 @@ var FieldImageUrl = FieldBinaryImage.extend({
...
                 var $button = $(".o_form_button_edit");
                 $button.openerpBounce();
                 e.stopPropagation();
             }
         });
         this.$('> img').remove();
         if (self.options.size) {
-            $img.css("width", "" + self.options.size[0] + "px");
-            $img.css("height", "" + self.options.size[1] + "px");
+            $img.css("width", "" + String(self.options.size[0]) + "px");
```
Comment (hparfr) [inline storage_image/static/src/js/storage_image.js]: not needed

File: storage_image/tests/test_storage_image.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _create_storage_image(self):
+        return self.env['storage.image'].create({
+            'name': self.filename,
+            'datas': self.filedata,
+            })
+
+    def test_create_and_read_image(self):
+        image = self._create_storage_image()
+        self.assertEqual(image.datas, self.filedata)
```
Comment (hparfr) [inline storage_image/tests/test_storage_image.py]: self.assertTrue(len(self.filedata)>1)) other whise "" == ""

File: storage_file/models/storage_file.py (hunk 12 above)
Comment (sebastienbeau) [reply storage_file/models/storage_file.py]: as discuses in the realworld for now this solution is enough and simple

File: storage_backend/models/storage_backend.py (hunk 1 above)
Comment (sebastienbeau) [reply storage_backend/models/storage_backend.py:23]: as discuses in the realworld for now this solution is enough and simple

**Pull request merged**